client/handle_http.go

```diff
@@ -1275,132 +1268,96 @@ func (tj *TransferJob) ID() string {
 		})
 		if isProxyEnabled && CanDisableProxy() {
 			details = append(details, transferAttemptDetails{
-				Url:        cacheURL,
+				Url:        remoteUrl,
 				Proxy:      false,
 				PackOption: opts.PackOption,
 			})
 		}
 	} else {
 		det := transferAttemptDetails{
-			Url:        cacheURL,
+			Url:        remoteUrl,
 			Proxy:      false,
 			PackOption: opts.PackOption,
 		}
-		if cacheURL.Scheme == "unix" {
-			det.UnixSocket = cacheURL.Path
+		if remoteUrl.Scheme == "unix" {
+			det.UnixSocket = remoteUrl.Path
 		}
 		details = append(details, det)
 	}
 
 	return details
 }
 
-type CacheInterface interface{}
-
-func generateTransferDetailsUsingCache(cache CacheInterface, opts transferDetailsOptions) []transferAttemptDetails {
-	if directorCache, ok := cache.(namespaces.DirectorCache); ok {
-		return newTransferDetailsUsingDirector(directorCache, opts)
-	} else if cache, ok := cache.(namespaces.Cache); ok {
-		return newTransferDetails(cache, opts)
-	}
-	return nil
-}
-
-// This function gets the amount of caches we will try equal to the configured "cachesToTry". It sets up our transfer attempt details for us and returns it.
-// This function also ensures that we do not try any duplicate caches
-func getCachesToTry(closestNamespaceCaches []CacheInterface, job *TransferJob, cachesToTry int, packOption string) (transfers []transferAttemptDetails) {
-	// The counter for the amount of caches we currently have listed to try
-	cachesListed := 0
-	// This list keeps track of the caches we have seen to ensure we do not have any duplicates
-	cacheList := make(map[CacheInterface]bool)
-	// Caches is just the list of caches we want to try (we print this out in debug logs)
-	caches := make([]CacheInterface, 0)
+// Generate the unique list of object servers (caches or origins) that will be attempted for a single transfer job and populate this info
+// in the slice of transferAttemptDetails structs
+func getObjectServersToTry(sortedObjectServers []string, job *TransferJob, oServersToTry int, packOption string) (transfers []transferAttemptDetails) {
+	oServersListed := 0
+	oServerList := make(map[string]bool)
+	oServers := make([]string, 0)
 
-	for _, cache := range closestNamespaceCaches {
-		// If we listed the amount of caches we want to try, we can end our loop
-		if cachesListed == cachesToTry {
+	for _, oServer := range sortedObjectServers {
+		if oServersListed == oServersToTry {
 			break
 		}
-		// if the current cache is already a cache we are trying, skip it
-		if cacheList[cache] {
+		// Handle deduplication of any object servers
+		if oServerList[oServer] {
 			continue
 		} else {
-			cachesListed++
-			caches = append(caches, cache)
-			cacheList[cache] = true
+			oServersListed++
+			oServers = append(oServers, oServer)
+			oServerList[oServer] = true
 			td := transferDetailsOptions{
-				NeedsToken: job.namespace.ReadHTTPS || job.namespace.UseTokenOnRead,
+				NeedsToken: !job.dirResp.XPelNsHdr.RequireToken,
```

Comment:
Shouldn't "NeedsToken" be true if RequireToken is True in the XPelNxHdr? Why the "!"?

Reply:
Good catch!